### PR TITLE
feat(start_planner): fix non-thread-safe access

### PR DESCRIPTION
**Pull request by soblin**
## Description

related work for https://github.com/autowarefoundation/autoware.universe/pull/6740

following variables are copied to thread local variable for thread-safety
- planner_data_
- getCurrentStatus() result
- parameters_
- isStopped() result
- `status_` (PullOutStatus)

## Related links

https://github.com/autowarefoundation/autoware.universe/issues/6718

## Tests performed

https://evaluation.tier4.jp/evaluation/reports/51d9494d-0cb2-5726-bd8e-80583125eb27?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [X] I've confirmed the [contribution guidelines].
- [X] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [X] The PR follows the [pull request guidelines].
- [X] The PR has been properly tested.
- [X] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [X] There are no open discussions or they are tracked via tickets.
- [X] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (danielsanchezaran) [inline planning/behavior_path_start_planner_module/src/start_planner_module.cpp:217]: The if and else are deeply nested into the brackets, which makes the code a little harder to read when the `if (freespace_thread_status_)` line is reached (its an if inside and else , inside a lock_guard set of brackets)
I suggest using a lambda function []() -> StartPlannerData{}
So you can have an early return, something like this:
```
...
 sp_planner_data.value().update(
        *parameters_, *planner_data_, getCurrentStatus(), status_, isStopped());
return sp_planner_data;
}
auto & sp_planner_data = sp_planner_data_.value();
...
```

and you can eliminate the "else"

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (danielsanchezaran) [inline planning/behavior_path_start_planner_module/src/start_planner_module.cpp:114]: This method is only called here, I suggest either modifying the method to accept is_stopped , pull_out_status.planner_type and pull_out_status.found_pull_out_path

and call
` isStuck( is_stopped , pull_out_status.planner_type ,pull_out_status.found_pull_out_path) `

Alternatively, the isStuck() method can be made into a lambda, inside this same function. 

It can also be deleted altogether, but I feel it is more readable if a method called is_ego_stuck() is called here instead of directly checking for 3 different  conditions on this if()

**Review by danielsanchezaran — COMMENTED** [review]
Thanks for this great PR, I have some minor comments regarding unused methods and variable names, but the changes themselves look solid, have you ran scenario tests with these changes applied?

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (danielsanchezaran) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp:217]: Also, since the update method is called either way: 

sp_planner_data.update(
        *parameters_, *planner_data_, getCurrentStatus(), 
Do we really need to check if sp_planner_data_ has a value? 
cannot we always do this:
`sp_planner_data_ = StartPlannerData();
 sp_planner_data_.value().update(
        *parameters_, *planner_data_, getCurrentStatus(), status_, isStopped());
`
?
Then you only need to check for `if (freespace_thread_status_)` ?

File: planning/behavior_path_start_planner_module/include/behavior_path_start_planner_module/start_planner_module.hpp
Comment (soblin) [reply planning/behavior_path_start_planner_module/include/behavior_path_start_planner_module/start_planner_module.hpp:173]: all right ! I'll replace it with "start_planner_data"

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (soblin) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp:1737]: I want to keep this method for now. As I'm doing in the body "begin/end of critical section", I'm creating a copy of StartPlannerData. Right now I'm making the copied of each member of StartPlannerData because most of the functions takes `const shared_ptr<PlannerData>` as the argument, not `const PlannerData&`, which is the reason why I'm making the copy of `StartPlannerData.planner_data` as the shared pointer.

In the future refactoring maybe I will replace most of the `const shared_ptr<PlannerData>` arguments to `const PlannerData&`, in which case the `clone()` method is useful.

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (soblin) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp:114]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/6779/commits/f12bd11c5b8c4b777bff93ef874d2069c0cc9495

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (soblin) [reply planning/behavior_path_start_planner_module/src/start_planner_module.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/6779/commits/95ed1a51e694db2a8ff153cec78cf66325e22723

I moved the checking of `freespace_thread_status_` out of the `else` scope.

I do not want to refresh `start_planner_data_` value every time due to following reason:
1. PlannerData has the shared pointer to RouteHandler, which contains the shared pointer to `lanelet_map_ptr` and `routing_graph_ptr`. I'm making the deep copy of `start_planner_data.planner_data.route_handler` so StartPlanner can become the owner (== increase reference count) of `lanelet_map_ptr` and `routing_graph_ptr`.
2. But if we use huge lanelet map, this refresh of RouteHandler is heavy because it contains the vector of lanelets. So it is preferable to refresh it only when needed, or when the map/route has changed so to speak.
3. currently PlannerData does not have the flag indicating if `the map/route has changed`, but I will add it in the next PR. With that flag, we can skip the refresh of `StartPlannerData.planner_data.route_handler` if the flag is false.

File: planning/behavior_path_start_planner_module/src/start_planner_module.cpp
Comment (danielsanchezaran) [inline planning/behavior_path_start_planner_module/src/start_planner_module.cpp:115]: Just a small detail, but could you please make the if(!local_planner_data || !current_status_opt || !parameters_opt || !pull_out_status_opt) condensed with a single boolean? 

like const bool is_data_ready = !local_planner_data || !current_status_opt || !parameters_opt || !pull_out_status_opt; 

In the future if  we need to check for more optional variables the if might become too large and CodeScene is complaining about this line.

**Review by danielsanchezaran — APPROVED** [review]
other than the multiple boolean comment I made everything else LGTM, I would appreciate it if @kyoichi-sugahara and @kosuke55 can also give their feedback before this PR is merged though. 

Thank you so much for this PR!

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!Thanks

**Pull request merged**
